### PR TITLE
Fix SQL when using single numeric with MySQL's json

### PR DESCRIPTION
**Pull request by razum2um**
### Summary

Currently when you insert a plain integer into a json field in mysql (which is a valid value according to rfc)
you cannot fetch this row by this value again. Happens only:
- on mysql, only with integers and floats, only inside json column: this is fixed
- on postgresql, only with integers and floats, only inside json column (**NOT** jsonb): workaround shown in tests

```ruby
numeric_value = 42
Model.create(json_column: numeric_value) # ok
Model.where(json_column: numeric_value)  # empty
```

### Current workaround

```ruby
Model.where("json_column = ?", numeric_value)  # ok + you have to rubocop:disable Rails/WhereEquals
```

### Tests

Added proper tests for all json-able databases, relevant runs for this PR
```sh
bundle exec rake test:sqlite3 TEST=test/cases/adapters/sqlite3/json_test.rb
bundle exec rake test:postgresql TEST=test/cases/adapters/postgresql/json_test.rb
bundle exec rake test:mysql2 TEST=test/cases/adapters/mysql2/json_test.rb
bundle exec rake test:mysql2 TEST=test/cases/json_attribute_test.rb
```

without this PR the newly added test fail like this:

```diff
cd activerecord; bundle exec rake test:mysql2 TEST=./test/cases/adapters/mysql2/json_test.rb

Failure:
Mysql2JSONTest#test_select_by_float_value [activerecord/test/cases/json_shared_test_cases.rb:224]:
--- expected
+++ actual
@@ -1 +1 @@
-#<JSONSharedTestCases::JsonDataType id: 1, payload: 1.234, settings: nil>
+nil

rails test activerecord/test/cases/json_shared_test_cases.rb:221

Failure:
Mysql2JSONTest#test_select_by_integer_value [activerecord/test/cases/json_shared_test_cases.rb:218]:
--- expected
+++ actual
@@ -1 +1 @@
-#<JSONSharedTestCases::JsonDataType id: 1, payload: 42, settings: nil>
+nil

rails test activerecord/test/cases/json_shared_test_cases.rb:215

27 runs, 55 assertions, 2 failures, 0 errors, 0 skips
```

<details>
<summary>Very detailed explanation what happens</summary>

MySQL adapter has no separate binds, it quotes values inline.
This affects case if (and only if) we pass explicit numeric as a value for `json` column in `where`

Note, the following queries are different

```sql
SELECT ... WHERE `json_data_type`.`payload` = 42                 -- a MySQL query in repl, finds correctly
SELECT ... WHERE `json_data_type`.`payload` = CAST('42' AS json) -- rendered NumericValue, finds correctly
```

But parent `Type::Json#serialize` returned json-string, it gets quoted
```sql
SELECT ... WHERE `json_data_type`.`payload` = '42' -- finds nothing
```

According to [RFC8259](https://datatracker.ietf.org/doc/html/rfc8259#section-3) a json top level value defined as:

> value = false / null / true / object / array / number / string

MySQL behaves correctly when making the difference

Besides, we cannot return numeric value from `#serialize`,
it breaks insertion, the following is invalid:

```sql
INSERT INTO `json_data_type` (`payload`) VALUES (42)
```

On the contrary, SQL part with `CAST` is fine
i.e. if (and only if) we pass explicit numeric as a value for `json` column in `create`, we'll get

```sql
INSERT INTO `json_data_type` (`payload`) VALUES (CAST('42' AS json))
```

To run a query that properly compares a `json` field with a single integer, which is valid operation,
we have to return an instance of this class and rely on `quoter` to call `to_s` on it.
It does the same with `when Numeric`, see `ActiveRecord::ConnectionAdapters::Mysql2Adapter#_quote(value)`
</details>

### CODE REVIEW

**Comment by p8** [comment]
Hey @razum2um 👋 Thanks for contributing to Rails!
I made some comments, but hopefully someone with more experience with json columns in mysql can have a look as well.

File: activerecord/test/cases/adapters/postgresql/json_test.rb [hunk 1]
```diff
@@ -41,6 +41,23 @@ class PostgresqlJSONTest < ActiveRecord::PostgreSQLTestCase
...
+    json = klass.create!(payload: 42)
+    x = klass.where("payload::text = ?::text", 42).first
+    assert_equal(json, x)
+  end
+
+  def test_select_by_float_value
+    json = klass.create!(payload: 1.234)
+    x = klass.where("payload::text = ?::text", 1.234).first
+    assert_equal(json, x)
+  end
```
[the remark this answers is not in the document]
Comment (razum2um) [reply activerecord/test/cases/adapters/postgresql/json_test.rb:60]: I suppose you meant adding test which passes `42` and `assert_raises`? ok, will do, but fwiw it's not just `ArgumentError ` but `ActiveRecord::StatementInvalid`

**Comment by razum2um** [comment]
Thanks for review @p8, fixed typos, added tests for PostgreSQL's json failure

File: activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb [hunk 2]
```diff
@@ -600,6 +601,13 @@ def initialize_type_map(m = type_map)
 
           m.register_type %r(^enum)i, Type.lookup(:string, adapter: :mysql2)
           m.register_type %r(^set)i,  Type.lookup(:string, adapter: :mysql2)
+          m.register_type(%r(^json)i) do |sql_type|
+            if /^json/.match?(sql_type)
+              MySQL::Json.new
```
Comment (zzak) [inline activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb:606]: Referencing the `MySQL` constant directly here seems like an anti-pattern to me, but I could definitely be wrong. Just by looking at the rest of this adapter anyways.

There is an explicit inclusion of [`MySQL::Quoting`](https://github.com/rails/rails/blob/e1cd663051b3b21511f7f5986cdca232aba6c136/activerecord/lib/active_record/connection_adapters/mysql/quoting.rb) and [`MySQL::SchemaStatements`](https://github.com/rails/rails/blob/e1cd663051b3b21511f7f5986cdca232aba6c136/activerecord/lib/active_record/connection_adapters/mysql/schema_statements.rb) though so maybe it's ok? I would have to look at the other adapters to feel more confident about this change.

Comment (razum2um) [reply activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb:606]: @zzak thanks for notice, but I thought about this: look at the `MySQL::SchemaStatements` file by the link you sent: there are `MySQL::TypeMetadata`, `MySQL::SchemaCreation `, `MySQL::SchemaDumper `, `MySQL::TableDefinition ` and  `MySQL::Column` constants (this choice is mostly influenced by it)

Comment (zzak) [reply activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb:606]: @razum2um Ok, but aren't those all constants from within Rails? Where does the `MySQL::Json` constant come from? 🤔 Sorry if this question is off track, I'm just trying to make sure I understand what is going on here.

Comment (razum2um) [reply activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb:606]: @zzak `MySQL::Json` is a new class in this PR :)

Comment (zzak) [reply activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb:606]: 🤦

**Comment by razum2um** [comment]
Solved a conflict, rebased. It's green. Anything I can do to improve it or to get another look?

**Pull request closed without merging**